src/runner_manager.py

```diff
@@ -0,0 +1,392 @@
+
+        try:
+            return next(
+                bin
+                for bin in runner_bins
+                if bin["os"] == os_name and bin["architecture"] == arch_name
+            )
+        except StopIteration as err:
+            raise RunnerBinaryError(
+                f"Unable query GitHub runner binary information for {os_name} {arch_name}"
+            ) from err
+
+    @retry(tries=5, delay=30, local_logger=logger)
+    def update_runner_bin(self, binary: RunnerApplication) -> None:
+        """Download a runner file, replacing the current copy.
+
+        Remove the existing runner binary to prevent it from being used. This
+        is done to prevent security issues arising from outdated runner binary
+        containing security flaws. The newest version of runner binary should
+        always be used.
+
+        Args:
+            binary: Information on the runner binary to download.
+        """
+        logger.info("Downloading runner binary from: %s", binary["download_url"])
+
+        # Delete old version of runner binary.
+        if self.runner_bin_path is not None:
+            self.runner_bin_path.unlink(missing_ok=True)
+            self.runner_bin_path = None
+
+        # Download the new file
+        response = self.session.get(binary["download_url"], stream=True)
+
+        logger.info(
+            "Download of runner binary from %s return status code: %i",
+            binary["download_url"],
+            response.status_code,
+        )
+
+        if not binary["sha256_checksum"]:
+            logger.error("Checksum for runner binary is not found, unable to verify download.")
+            raise RunnerBinaryError("Checksum for runner binary is not found in GitHub response.")
+
+        sha256 = hashlib.sha256()
+
+        # Use tempfile as recommend by bandit to avoid hardcoding the file
+        # path.  The `delete=False` arg save the file permanently for the charm
+        # to use start runners.
+        with tempfile.NamedTemporaryFile(delete=False) as tmp_file:
+            for chunk in response.iter_content(decode_unicode=False):
+                tmp_file.write(chunk)
+
+                sha256.update(chunk)
+
+        logger.info("Finished download of runner binary.")
+
+        # Verify the checksum if checksum is present.
+        if binary["sha256_checksum"] != sha256.hexdigest():
+            logger.error(
+                "Expected hash of runner binary (%s) doesn't match the calculated hash (%s)",
+                binary["sha256_checksum"],
+                sha256,
+            )
+            raise RunnerBinaryError("Checksum mismatch for downloaded runner binary")
+
+        # Verify the file integrity.
+        if not tarfile.is_tarfile(tmp_file.name):
+            logger.error("Failed to decompress downloaded GitHub runner binary.")
+            raise RunnerBinaryError("Downloaded runner binary cannot be decompressed.")
```

Comment:
Using the [`atexit`](https://docs.python.org/3/library/atexit.html) module to register a cleanup function in case of exception?

Reply:
There is indeed resource leak (tmp files) on file integrity check failure.
I am working on fixing it.